### PR TITLE
Merge editable mappings

### LINKED ISSUE #6955: Support merging of proofreading annotations
Follow-up to #6807. See design doc (todo: add link here).

Merging editable mappings should be defined as applying all actions of mapping B *after* all actions of mapping A. Instead of the agglomerate IDs in the update actions, only the positions should be used and the current state of the mappings as well as the oversegmentation should be used to look up both segment ids and agglomerate ids at those positions.

### LINKED ISSUE #7015: Clarify Merge Annotation Modal
By default, the checkbox is active, but this only treats trees. I think it should be unchecked by default, otherwise users won’t find a way to e merge volume annotations.

Also, wording could probably be improved to be more explanatory.

image

Related: #6088

### PR DESCRIPTION
**Pull request by fm3**
- Adds merge functionality for editable mapping (a.k.a. supervoxel proofreading) volume layers
- Fixes a bug in duplicating such annotations
- Adds small refactoring `Instant.since`
- Changes the merge modal from a checkbox to individual buttons to emphasize the difference between the options:
image

### TODO
 - [x] Depends on https://github.com/scalableminds/fossildb/pull/38 – after that release, change fossilDB version here and update migration guide

### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- While the new fossilDB is not released, this can be tested by checking out the FossilDB branch `fix-get-multiple-keys-pagination`, running `sbt "run --columnFamilies skeletons,skeletonUpdates,volumeData,volumes,volumeUpdates,editableMappings,editableMappingUpdates,editableMappingsInfo,editableMappingsAgglomerateToGraph,editableMappingsSegmentToAgglomerate"` there and `yarn start noFossilDB` in webknossos.
- Create two explorative editable mapping annotations on the same dataset, merge/split some agglomerates
- Use the merge modal to merge one with the other
- Result should have updates of both mappings applied
- Merge modal should be more understandable as well (feedback welcome)

### Issues:
- fixes #6955 
- fixes #7015

------
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

### CODE REVIEW

File: MIGRATIONS.unreleased.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ User-facing changes are documented in the [changelog](CHANGELOG.released.md).
...
+- For instances with existing editable mapping (a.k.a supervoxel proofreading) annotations: To keep those annotations alive, a python migration has to be run with access to your tracingstore’s FossilDB. It is recommended to do this during a webknossos downtime to avoid data loss. It needs python 3.8+ and the pip packages installable by `pip install grpcio-tools grpcio-health-checking`. Run it with `python tools/migrate-editable-mappings/migrate-editable-mappings.py -v -w -o localhost:7155`. Omit -o for a faster migration but no access to older versions of the editable mappings. The migration is idempotent.
+- A new FossilDB Version is required [TODO: Insert release number an link here once fossilDB PR https://github.com/scalableminds/fossildb/pull/38 is merged]
```
Comment (philippotto) [inline MIGRATIONS.unreleased.md:12]: todo

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 2]
```diff
@@ -141,40 +101,51 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
-      }
+      const annotation = await getAnnotationCompoundInformation(
+        selectedProject,
+        APIAnnotationTypeEnum.CompoundProject,
+      );
+      this.mergeAnnotationIntoActiveTracing(annotation);
+    }
+  };
+
+  handleImportProject = async (event: React.SyntheticEvent) => {
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:117]: looks pretty identical to `handleMergeProject`? did you delete the wrong branch in the if-clause?

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 3]
```diff
@@ -141,40 +101,51 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+
+  handleMergeExplorativeAnnotation = async (event: React.SyntheticEvent) => {
+    event.preventDefault();
+    const { selectedExplorativeAnnotation } = this.state;
+
+    if (selectedExplorativeAnnotation != null) {
+      const url =
+        `/api/annotations/Explorational/${selectedExplorativeAnnotation}/merge/` +
+        `${this.props.annotationType}/${this.props.annotationId}`;
+      this.merge(url);
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:138]: how about renaming the merge-method to `createMergedAnnotation` or something like that to make its behavior clearer?

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 4]
```diff
@@ -267,17 +226,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
               />
             </Form.Item>
-
-            <ButtonWithCheckbox
-              checkboxContent={mergeIntoActiveTracingCheckbox}
-              button={
-                // @ts-expect-error ts-migrate(2322) FIXME: Type '"default"' is not assignable to type 'SizeTy... Remove this comment to see the full error message
-                <Button type="primary" size="default" disabled={this.state.selectedProject == null}>
+            <Form.Item>
+              <Tooltip title="Imports trees and tree groups directly into the currently opened annotation.">
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:230]:
```suggestion
              <Tooltip title="Imports trees and tree groups (no volume data!) directly into the currently opened annotation.">
```

or is this too much?

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 5]
```diff
@@ -267,17 +226,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+                <Button
+                  type="primary"
+                  disabled={this.state.selectedProject == null}
+                  onClick={this.handleImportProject}
+                >
+                  Import trees here
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:236]:
```suggestion
                  Import trees into current annotation
```

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 6]
```diff
@@ -267,17 +226,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+              </Tooltip>
+            </Form.Item>
+            <Form.Item>
+              <Tooltip title="Creates a new explorative annotation in your account with all merged contents of the current and selected annotations.">
+                <Button
+                  type="primary"
+                  disabled={this.state.selectedProject == null}
+                  onClick={this.handleMergeProject}
+                >
                   Merge
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:247]:
```suggestion
                  Merge into new annotation
```

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 7]
```diff
@@ -290,20 +260,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+            <Form.Item>
+              <Tooltip title="Imports trees and tree groups directly into the currently opened annotation.">
                 <Button
                   type="primary"
-                  // @ts-expect-error ts-migrate(2322) FIXME: Type '"default"' is not assignable to type 'SizeTy... Remove this comment to see the full error message
-                  size="default"
                   disabled={this.state.selectedExplorativeAnnotation.length !== 24}
+                  onClick={this.handleImportExplorativeAnnotation}
+                >
+                  Import trees here
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:270]:
```suggestion
                  Import trees into current annotation
```

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 8]
```diff
@@ -290,20 +260,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+                </Button>
+              </Tooltip>
+            </Form.Item>
+            <Form.Item>
+              <Tooltip title="Creates a new explorative annotation in your account with all merged contents of the current and selected annotations.">
+                <Button
+                  type="primary"
+                  disabled={this.state.selectedExplorativeAnnotation.length !== 24}
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:278]: I think, this line is incorrect :sweat_smile: I tracked it down to https://github.com/scalableminds/webknossos/commit/66527b720af208906ecb9f7e76f99626d5558bed, where it didn't really make sense, either. Probably length should be compared against 0?

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx [hunk 9]
```diff
@@ -290,20 +260,28 @@ class _MergeModalView extends PureComponent<Props, MergeModalViewState> {
...
+                  onClick={this.handleMergeExplorativeAnnotation}
                 >
                   Merge
```
Comment (philippotto) [inline frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:281]:
```suggestion
                  Merge into new annotation
```

**Review by philippotto — COMMENTED** [review]
awesome :+1: I left some wording suggestions, but they might be too verbose or too on the nose. however, I still think, there is room to make the UI clearer. let me know what you think.

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx (hunk 2 above)
Comment (fm3) [reply frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:116]: whoops, so I did! Good catch

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx (hunk 8 above)
Comment (philippotto) [reply frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:277]: Ohhh, I see. The random 24 caught me off-guard and I couldn't make sense of it. how about extracting it into a `ANNOTATION_ID_LENGTH` constant?

Comment (philippotto) [reply frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:277]: sorry for the confusion!

**Comment by fm3** [comment]
I’d prefer not to change the text on the buttons. I had longer variants at first, but it made for a very cluttered UI and very large buttons and line breaks on small screens. I think with the tooltips, this should be ok. What do you think?

File: frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx (hunk 8 above)
Comment (fm3) [reply frontend/javascripts/oxalis/view/action-bar/merge_modal_view.tsx:277]: No problem :) Sure, constant sounds good. Where would I put that? I’m always a bit lost in the tsx files to be honest :sweat_smile:

**Review by philippotto — COMMENTED** [review]
front-end looks good to me :+1: i'm leaving the final acceptance to the backend review @frcroth

File: webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/editablemapping/EditableMappingService.scala [hunk 10]
```diff
@@ -588,4 +629,55 @@ class EditableMappingService @Inject()(
...
+    for {
+      firstMappingId <- editableMappingIds.headOption.toFox
+      before = Instant.now
+      newMappingId <- duplicate(Some(firstMappingId), version = None, remoteFallbackLayer, userToken)
+      _ <- Fox.serialCombined(editableMappingIds.tail)(editableMappingId =>
+        mergeInto(newMappingId, editableMappingId, remoteFallbackLayer, userToken))
+      _ = logger.info(s"Merging ${editableMappingIds.length} editable mappings took ${Instant.since(before)}")
+    } yield newMappingId
+
+  private def mergeInto(targetEditableMappingId: String,
```
Comment (frcroth) [inline webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/editablemapping/EditableMappingService.scala:644]: Add comment for the meaning of source and target

File: webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/editablemapping/EditableMappingService.scala [hunk 11]
```diff
@@ -588,4 +629,55 @@ class EditableMappingService @Inject()(
...
+                        sourceEditableMappingId: String,
+                        remoteFallbackLayer: RemoteFallbackLayer,
+                        userToken: Option[String]): Fox[Unit] =
+    for {
+      targetNewestVersion <- getClosestMaterializableVersionOrZero(targetEditableMappingId, None)
+      closestMaterializedWithVersion <- getInfoAndActualVersion(sourceEditableMappingId,
```
Comment (frcroth) [inline webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/editablemapping/EditableMappingService.scala:650]:
```suggestion
      newestMaterializedWithVersion <- getInfoAndActualVersion(sourceEditableMappingId,
```

**Review by frcroth — APPROVED** [review]

**Pull request merged**